In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

plt.rc('font', family='Malgun Gothic')
plt.rcParams['axes.unicode_minus'] = False

# 머신러닝

## 1 주택 계약 데이터

In [4]:
df = pd.read_csv(
    'https://raw.githubusercontent.com/ADPclass/ADP_book_ver01/main/data/29_problem1.csv',
encoding='cp949')
df.head()

순번 계약구분  재계약횟수  거주개월 아파트 이름  아파트 ID  아파트 평점  호실고유번호  층  평형대  ...  퇴거연도  \
0   1   유효     10   222  강남아파트       5     7.0   14520  1   12  ...   NaN   
1   1   유효     10   222  강남아파트       5     7.0   14520  1   12  ...   NaN   
2   1   유효     10   222  강남아파트       5     7.0   14520  1   12  ...   NaN   
3   1   유효     10   222  강남아파트       5     7.0   14520  1   12  ...   NaN   
4   1   유효     10   222  강남아파트       5     7.0   14520  1   12  ...   NaN   

   거주연도  월세(원)   보증금(원)  대표나이  나이  성별  결혼여부  거주자 수 퇴거여부  
0  2008  47100  3646000    46  33   남    미혼      3  미퇴거  
1  2009  56500  4375000    46  34   남    미혼      3  미퇴거  
2  2010  56500  4375000    46  35   남    미혼      3  미퇴거  
3  2011  69900  5408000    46  36   남    미혼      3  미퇴거  
4  2012  69900  5408000    46  37   남    미혼      3  미퇴거  

[5 rows x 23 columns]

- (1) 계약자 고유번호(contract_id) 기준으로 가장 최신의  
  거주연도(residence_year)만 포함하는 데이터프레임 생성
- (2) (1)의 데이터로 EDA 진행 및 결측치 처리
- (3) (2)의 데이터에서 이상치 처리
- (4) (3)의 데이터에서 재계약 횟수(renewal_counts)의 중앙값을 기준으로  
  중앙값보다 크거나 같으면 'High', 작으면 'Low'로 renewal_h/l 변수 생성
- (5) (4)의 데이터로 차원을 축소하고 필요한 이유 기술.  
  필요하지 않다면 하지 않고 그 이유를 기술
- (6) (4) 데이터의 renewal_h/l 변수를 기준으로 데이터를 나누고,  
  각 데이터의 특징을 서술
- (7) 재계약횟수를 종속변수로하는 회귀분석을 두 가지 이상의 방법론을  
  통해 수행하고, 최종 모델을 결정
- (8) 최종 모델에서 유의하게 작용하는 변수를 확인하고 설명
- (9) 해당 분석결과로 얻을 수 있는 인사이트 제시

## 2 야구 데이터

In [5]:
df = pd.read_csv(
    'https://raw.githubusercontent.com/ADPclass/ADP_book_ver01/main/data/29_problem2.csv')
df.head()

game_id  a1_1  a1_2  a2_1  a2_2  a3_1  a3_2  a4_1  a4_2  a5_1  ...  a9_2  \
0  201900016     5     5     5     5     5     5     5     5     5  ...     5   
1  201900023     6     4     5     5     2     5     1     1     6  ...     7   
2  201900103     5     6     5     5     1     9     5     4     6  ...     6   
3  201900112     5     7     6     1     5     5     1     5     1  ...     5   
4  201900131     5     1     2     5     2     5     7     2     6  ...     6   

   b1  b2  b3  b4  b5  b6  b7  b8  b9  
0   0   0   0   0   0   0   0   4   2  
1   3   0   0   1   0   1   3   1   0  
2   0   0   1   1   4   1   2   0   1  
3   0   3   0   0   0   0   1   0   0  
4   0   2   4   1   2   2   0   1   0  

[5 rows x 28 columns]

- (1) 회차별로 1번 타자의 출루 (1 ~ 3루타, 사사구) 가 있는  
  경우에 대해 득점이 발생했는지 확인하는 분석을 위한 전처리 수행  
  (단 첫 번째 혹은 두 번째 타자가 홈런을 친 경우 해당 회차 데이터는 제외)
- (2) (1)에 데이터에 대해 Logistic Regression을 적용하고,  
  2번 타자의 희생번트 여부에 대한 회귀계수 검정을 적용
- (3) SMOTE를 적용하여 데이터 불균형을 해결하시오
- (4) (3)의 데이터에 Logistic Regression을 적용하고 결과 분석

# 통계분석

## 1

제품 A의 불량률은 0.03이다. 25개의 제품을 뽑았을 때  
3개가 불량일 확률은? (소수점 다섯째 자리에서 반올림)

## 2 데이터 분석 후 기술

- (1) C사 생산제품 1000개 중 양품이 600개, D사 생산제품 500개 중 양품이 200개이다.  
  두 회사의 양품률에 차이가 있는지 검정
- (2) 각 차종별 범퍼 파손의 정도에 차이가 유의한지 검정
- (3) 귀무가설 채택 시 그 의미를 해석하고, 기각이라면 사후분석 시행

### 2-1

### 2-2

In [7]:
data = pd.read_csv(
    'https://raw.githubusercontent.com/ADPclass/ADP_book_ver01/main/data/29_problem3.csv')
data.head()

name      ratio
0    A  10.993428
1    A   9.723471
2    A  11.295377
3    A  13.046060
4    A   9.531693

### 2-3

## 3

L1 ~ L3 세 개의 생산라인에서 각각 13%, 37%, 50%를 생산하며,  
각각 1.1%, 2.1%, 3.3%의 불량률을 갖는다.  
불량 제품이 나왔을 때 L1 라인에서 생산되었을 확률은?  
(소수점 둘째 자리에서 반올림)